### PR TITLE
Sqlite: Support spatial data via NTS and SpatiaLite

### PR DESCRIPTION
**Pull request by bricelam**
Usage is pretty much the same as SQL Server:

``` csharp
optionsBuilder.UseSqlite(
    connectionString,
    x => x.UseNetTopologySuite())
```

``` csharp
class SpatialThing
{
    public int Id { get; set; }
    public IPoint Point { get; set; }
}
```

``` csharp
var currentLocation = new Point(-122.13, 47.64);
var nearestThings = from t in db.SpatialThings
                    where t.Point.Distance(currentLocation) < 1500 // meters
                    select t;
```

However, on SpatiaLite, the SRID (and dimension) needs to be specified on the column:

``` csharp
modelBuilder.Entity<SpatialThing>()
    .Property(t => t.Point)
        .ForSqliteHasSrid(4326);
```

The `mod_spatialite` NuGet package includes the native dependencies for Windows x86 and x64, but other runtimes will need to install it globally. For example:

``` sh
# Ubuntu
sudo apt-get install -y libsqlite3-mod-spatialite

# macOS
brew install libspatialite
```

Part of #1100 

See also:
* aspnet/Universe#1389

### CODE REVIEW

File: src/EFCore.Sqlite.NTS/EFCore.Sqlite.NTS.csproj [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\EFCore.Sqlite.Core\EFCore.Sqlite.Core.csproj" PrivateAssets="contentfiles;build" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="mod_spatialite" Version="$(mod_spatialitePackageVersion)" />
+    <PackageReference Include="NetTopologySuite.Core" Version="$(NetTopologySuiteCorePackageVersion)" />
+    <PackageReference Include="NetTopologySuite.IO.SpatiaLite" Version="$(NetTopologySuiteIOSpatiaLitePackageVersion)" />
```
Comment (bricelam) [inline src/EFCore.Sqlite.NTS/EFCore.Sqlite.NTS.csproj]: @Eilon Do we need additional approval for this NTS package?

File: src/EFCore.Specification.Tests/SharedStoreFixtureBase.cs [hunk 2]
```diff
@@ -71,9 +71,13 @@ public DbContextOptions CreateOptions()
             => ConfigureOptions(ServiceProvider, new DbContextOptionsBuilder()).Options;
 
         private DbContextOptionsBuilder ConfigureOptions(IServiceProvider serviceProvider, DbContextOptionsBuilder optionsBuilder)
-            => AddOptions(TestStore.AddProviderOptions(optionsBuilder))
+            => AddOptions(TestStore.AddProviderOptions(optionsBuilder, x => ConfigureProviderOptions(x)))
                 .UseInternalServiceProvider(serviceProvider);
 
+        protected virtual void ConfigureProviderOptions(dynamic dynamicOptionsBuilder)
```
Comment (AndriySvyryd) [inline src/EFCore.Specification.Tests/SharedStoreFixtureBase.cs]: There is no need to add this. Just create a new provider-specific options builder in the provider-specific fixture, see NorthwindRowNumberPagingQuerySqlServerFixture

Comment (bricelam) [reply src/EFCore.Specification.Tests/SharedStoreFixtureBase.cs]: Ah thanks, I had a feeling I was over-complicating things.

File: test/EFCore.Sqlite.FunctionalTests/TestUtilities/SpatialiteRequiredAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+                        return SpatialiteLoader.TryLoad(connection);
+                    }
+                });
+
+        public bool IsMet
+            => _loaded.Value;
+
+        public string SkipReason
+            => "mod_spatialite not found. Install it to run this test.";
```
Comment (bricelam) [inline test/EFCore.Sqlite.FunctionalTests/TestUtilities/SpatialiteRequiredAttribute.cs]: @natemcmaster Any way to do the following on the agents?

``` sh
# linux
sudo apt-get install -y libsqlite3-mod-spatialite

# osx
brew install libspatialite
```

File: test/EFCore.Sqlite.FunctionalTests/SpatialSqliteTest.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.EntityFrameworkCore.TestUtilities;
+
+namespace Microsoft.EntityFrameworkCore
+{
+    [SpatialiteRequired]
+    public class SpatialSqliteTest : SpatialTestBase<SpatialSqliteFixture>
```
Comment (AndriySvyryd) [inline test/EFCore.Sqlite.FunctionalTests/SpatialSqliteTest.cs]: Add #if !Test21

File: test/EFCore.Sqlite.FunctionalTests/SpatialSqliteFixture.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.EntityFrameworkCore
+{
+    public class SpatialSqliteFixture : SpatialFixtureBase
```
Comment (AndriySvyryd) [inline test/EFCore.Sqlite.FunctionalTests/SpatialSqliteFixture.cs]: Add #if !Test21

File: test/EFCore.Sqlite.FunctionalTests/Query/SpatialQuerySqliteFixture.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
```
Comment (AndriySvyryd) [inline test/EFCore.Sqlite.FunctionalTests/Query/SpatialQuerySqliteFixture.cs]: Add #if !Test21

File: test/EFCore.Sqlite.FunctionalTests/Query/SpatialQuerySqliteTest.cs [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Xunit.Abstractions;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    [SpatialiteRequired]
+    public class SpatialQuerySqliteTest : SpatialQueryTestBase<SpatialQuerySqliteFixture>
```
Comment (AndriySvyryd) [inline test/EFCore.Sqlite.FunctionalTests/Query/SpatialQuerySqliteTest.cs]: Add #if !Test21

File: src/EFCore.Sqlite.NTS/EFCore.Sqlite.NTS.csproj (hunk 1 above)
Comment (Eilon) [reply src/EFCore.Sqlite.NTS/EFCore.Sqlite.NTS.csproj]: Yup, every version of every package, that way we get notifications in the future if there's a specific issue with the specific instance.

I registered this one.

BTW new automated registration is coming soon. That way when I'm out sick (like last Friday) there will be a robot to do my job 😄

**Pull request merged**